File: src/main/java/com/ada/conversation/context/SystemContextItem.java
package com.ada.conversation.context;

import com.ada.conversation.application.SystemPromptProvider;
import com.ada.conversation.application.dto.*;
import com.ada.shared.observability.MeasuredContextItem;
import lombok.RequiredArgsConstructor;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

@Component
@Order(10)
@MeasuredContextItem("system")
@RequiredArgsConstructor
public class SystemContextItem implements ContextItem {
  private final SystemPromptProvider provider;

  public LlmContentComponent component() {
    return LlmContentComponent.SYSTEM;
  }

  public ContextState apply(ChatRequest r, ContextState c) {
    var m = new java.util.ArrayList<>(c.messages());
    m.add(new LlmMessage(LlmMessageRole.SYSTEM, provider.content(), component()));
    return new ContextState(m, c.tools(), c.selection());
  }
}

File: src/main/kotlin/com/ada/conversation/context/SystemContextItem.kt
package com.ada.conversation.context

import com.ada.conversation.application.SystemPromptProvider
import com.ada.conversation.application.dto.ChatRequest
import com.ada.conversation.application.dto.ContextState
import com.ada.conversation.application.dto.LlmContentComponent
import com.ada.conversation.application.dto.LlmMessage
import com.ada.conversation.application.dto.LlmMessageRole
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import com.ada.shared.observability.MeasuredContextItem

@Component
@Order(10)
@MeasuredContextItem("system")
class SystemContextItem(
    private val systemPromptProvider: SystemPromptProvider,
) : ContextItem {
    override val component = LlmContentComponent.SYSTEM

    override fun apply(request: ChatRequest, current: ContextState): ContextState = current.copy(
        messages = current.messages + LlmMessage(LlmMessageRole.SYSTEM, systemPromptProvider.content(), component),
    )
}
